### ISSUE TITLE
Bug - Programatically Refreshing HTML Breaks AMP Bindings (amp-state)

### ISSUE THREAD

**Issue by stephengardner** [ampproject/amphtml#27918]
Previously verified by @nainar, hoping to get some help here.

Full demo Stackblitz: https://stackblitz.com/edit/shopsheriff-amp-bind-error?file=index.html

When creating AMP pages using a dynamic page-builder, such as the editor we've built for AMP on Shopify, dynamic HTML injection typically updates a page.

When doing so, AMP state values break.  All elements that rely on state become static.

The Stackblitz has two complete examples of this behaviour.

**Comment by nainar** [comment]
@ampproject/wg-runtime to take a look

**Comment by samouri** [comment]
That makes sense. `<amp-bind>` doesn't automatically rescan the page for bindings when the document has its contents replaced.

There are a couple of ways I could see us fixing this:
1. Have a listener that rescans for bindings on any non-AMP caused dom mutation. This could happen automatically, but I'd be worried it could impede performance.
2. Expose a JS function on the `AMP` global for developers to call if they need to retrigger a scan.

**Comment by stephengardner** [comment]
Thanks @nainar and @samouri 

#2 seems like the quickest and least troubling solution.  We could certainly call a function to re-bind state, that'd be easy peasy.

We were looking for any kind of a hacky way around this in the interim, like re-appending the amp runtime script to the head of an iframe, but to no avail.  We'd love to have any solution and we're comfortable with calling functions/performing actions on our end as well.

Thanks again

**Comment by samouri** [comment]
The hackiest thing I could imagine that may work is writing something like this:
```js
    window.__AMP_SERVICES.bind.obj.rescan(
      /* added elements */ [document.getElementById('contents')], 
      /* removed elements*/ [document.getElementById('contents')], 
      {update: false }
    );
```

That said, I _really_ wouldn't recommend it. It isn't part of our public API and will likely break soon.

From this example, it looks like that snippet fixes bind, but there are still issues with the carousel that may need to separately be addressed.

**Comment by stephengardner** [comment]
Thanks @samouri 
While looking into that hack, I noticed other components are also un-bound, such as "amp-fx-collection" for a parallax effect.  Once HTML is replaced, the parallax effect is nullified.  Though the hack does work for state.  
We'll heed your advice and not rely on the hack.
This would be a huge help if something like this were created to re-bind components.  I can't imagine we're the only page-builder that refreshes HTML content and would benefit from this.  But we have thousands of ecommerce brands using the page builder so we'd be happy to help in any way/shape/form.
Looking forward to an update

**Comment by choumx** [comment]
This is not supported yet. WordPress's AMP plugin had similar problems and this "Bento" initiative was designed in part to address this general "reactivity" issue. The only real solution today is to reload your rendered preview after markup changes, similar to what JSBin et al do.

/cc @westonruter @dvoytenko

**Comment by stephengardner** [comment]
@choumx Yes, we're reloading the preview as an interim approach.  This is unfortunate of course for obvious UX reasons, so looking forward to a way to re-bind state programmatically.

**Comment by stephengardner** [comment]
Just wanted to follow up here and see if any progress has been made, or if there's additional intent to progress here.  

We've  also noticed that things like <amp-user-notification> could benefit from this re-binding of the AMP javascript, so that a full page refresh doesn't need to take place when adding components via a page-builder type setup.  Currently, injecting an <amp-user-notification> element would just show up blank.

**Comment by dvoytenko** [comment]
@stephengardner There's quite a big of progress has been made here. Although, `amp-user-notification` hasn't yet been looked at. You can follow the updates on [wg-bento](https://github.com/ampproject/wg-bento) working group. At the next AMP Dev Summit, there should be a session on the progress of this project as well.

/cc @nainar

**Comment by stephengardner** [comment]
Thanks @dvoytenko, I'm confused as to what progress has been made, since we're still seeing the issue occurring regarding amp-state binding.  Is there now a way to programmatically rebind state?  Is there a separate public issue addressing this?  Is there a solution, or just movement towards a solution in the future?
Thanks

**Comment by choumx** [comment]
Short-term, you can consider hooking into a private API by dispatching an event with type `'amp:dom-update'` where the target is the closest ancestor of the changed element(s). 

https://github.com/ampproject/amphtml/blob/919f1640be9bebda89838edff4834d4bfd225a86/extensions/amp-bind/0.1/bind-impl.js#L626-L630

Note that this is not an official API and is subject to change (though it has been stable in recent history).

**Comment by samouri** [comment]
@stephengardner, I'm surprised the fix mentioned by choumx didn't work. I'm out for the next few weeks but will investigate in early June when I'm back
